File: services/graphql-tms/src/mutation/carriers.rs
use async_graphql::{Context, InputObject, Object};
use rust_decimal::Decimal;
use uuid::Uuid;

use crate::models::{carrier_rates, carriers, sea_orm_active_enums::CarrierRateUnitEnum};

#[derive(Debug, Clone, InputObject)]
pub struct CreateCarrierRateInput {
    pub service_type: Option<String>,
    pub origin: Option<String>,
    pub destination: Option<String>,
    pub rate: Decimal,
    pub unit: Option<CarrierRateUnitEnum>,
}

#[derive(Debug, Clone, InputObject)]
pub struct CreateCarrierInput {
    pub name: String,
    pub contact_details: Option<String>,
    pub services_offered: Option<String>,
    pub rates: Vec<CreateCarrierRateInput>,
}

#[derive(Debug, Clone, Default)]
pub struct Mutation;

#[Object(name = "TmsCarriersMutation")]
impl Mutation {
    async fn create_carrier(
        &self,
        ctx: &Context<'_>,
        payload: CreateCarrierInput,
    ) -> async_graphql::Result<carriers::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let mut trx = db.begin().await?;

        let result = sqlx::query_as::<_, carriers::Model>(
            "insert into tms.carriers (name, contact_details, services_offered) values ($1, $2, $3) returning *",
        )
        .bind(payload.name)
        .bind(payload.contact_details)
        .bind(payload.services_offered)
        .fetch_one(&mut *trx)
        .await?;

        for rate in payload.rates {
            _ = sqlx::query("insert into tms.carrier_rates (carrier_id, service_type, origin, destination, rate, unit) values ($1, $2, $3, $4, $5, $6)")
                .bind(result.id)
                .bind(rate.service_type)
                .bind(rate.origin)
                .bind(rate.destination)
                .bind(rate.rate)
                .bind(rate.unit)
                .execute(&mut *trx)
                .await?;
        }

        trx.commit().await?;

        Ok(result)
    }

    async fn update_carrier_name(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        name: String,
    ) -> async_graphql::Result<carriers::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(
            sqlx::query_as("update tms.carriers set name = $1 where id = $2 returning *")
                .bind(name)
                .bind(id)
                .fetch_one(db)
                .await?,
        )
    }

    async fn update_carrier_contact_details(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        contact_details: String,
    ) -> async_graphql::Result<carriers::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(
            sqlx::query_as(
                "update tms.carriers set contact_details = $1 where id = $2 returning *",
            )
            .bind(contact_details)
            .bind(id)
            .fetch_one(db)
            .await?,
        )
    }

    async fn update_carrier_services_offered(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        services_offered: String,
    ) -> async_graphql::Result<carriers::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as(
            "update tms.carriers set services_offered = $1 where id = $2 returning *",
        )
        .bind(services_offered)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn remove_carrier(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from tms.carriers where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() == 0 {
            return Err("Unable to remove carrier".into());
        }

        Ok("Carrier removed".into())
    }

    async fn add_carrier_rate(
        &self,
        ctx: &Context<'_>,
        carrier_id: Uuid,
        payload: CreateCarrierRateInput,
    ) -> async_graphql::Result<carriers::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        _ = sqlx::query("insert into tms.carrier_rates (carrier_id, service_type, origin, destination, rate, unit) values ($1, $2, $3, $4, $5, $6)")
            .bind(carrier_id)
            .bind(payload.service_type)
            .bind(payload.origin)
            .bind(payload.destination)
            .bind(payload.rate)
            .bind(payload.unit)
            .execute(db)
            .await?;

        Ok(sqlx::query_as("select * from tms.carriers where id = $1")
            .bind(carrier_id)
            .fetch_one(db)
            .await?)
    }

    async fn update_carrier_rate(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        payload: CreateCarrierRateInput,
    ) -> async_graphql::Result<carrier_rates::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as("update tms.carrier_rates set service_type = $1, origin = $2, destination = $3, rate = $4, unit = $5 where id = $6 returning *")
            .bind(payload.service_type)
            .bind(payload.origin)
            .bind(payload.destination)
            .bind(payload.rate)
            .bind(payload.unit)
            .bind(id)
            .fetch_one(db)
            .await?)
    }

    async fn remove_carrier_rate(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
    ) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from tms.carrier_rates where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() == 0 {
            return Err("Unable to remove carrier rate".into());
        }

        Ok("Carrier rate removed".into())
    }
}

File: src/graphql/mutations/tms/carriers.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// CARRIER MUTATIONS
// ============================================================================

/**
 * Creates a new carrier.
 * @param payload The input data for creating the carrier.
 */
export const createCarrier = graphql(`
  mutation CreateCarrier($payload: CreateCarrierInput!) {
    tms {
      createCarrier(payload: $payload) {
        id
        name
        contactDetails
        servicesOffered
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Updates the name of a carrier.
 * @param id The UUID of the carrier to update.
 * @param name The new name.
 */
export const updateCarrierName = graphql(`
  mutation UpdateCarrierName($id: UUID!, $name: String!) {
    tms {
      updateCarrierName(id: $id, name: $name) {
        id
        name
        contactDetails
        servicesOffered
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Updates the contact details of a carrier.
 * @param id The UUID of the carrier to update.
 * @param contactDetails The new contact details.
 */
export const updateCarrierContactDetails = graphql(`
  mutation UpdateCarrierContactDetails($id: UUID!, $contactDetails: String!) {
    tms {
      updateCarrierContactDetails(id: $id, contactDetails: $contactDetails) {
        id
        name
        contactDetails
        servicesOffered
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Updates the services offered by a carrier.
 * @param id The UUID of the carrier to update.
 * @param servicesOffered The new services offered.
 */
export const updateCarrierServicesOffered = graphql(`
  mutation UpdateCarrierServicesOffered($id: UUID!, $servicesOffered: String!) {
    tms {
      updateCarrierServicesOffered(id: $id, servicesOffered: $servicesOffered) {
        id
        name
        contactDetails
        servicesOffered
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Removes a carrier by its ID.
 * @param id The UUID of the carrier to remove.
 */
export const removeCarrier = graphql(`
  mutation RemoveCarrier($id: UUID!) {
    tms {
      removeCarrier(id: $id)
    }
  }
`);

/**
 * Adds a new carrier rate to a carrier.
 * @param carrierId The UUID of the carrier to add the rate to.
 * @param payload The input data for creating the carrier rate.
 */
export const addCarrierRate = graphql(`
  mutation AddCarrierRate($carrierId: UUID!, $payload: CreateCarrierRateInput!) {
    tms {
      addCarrierRate(carrierId: $carrierId, payload: $payload) {
        id
        name
        contactDetails
        servicesOffered
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Updates an existing carrier rate.
 * @param id The UUID of the carrier rate to update.
 * @param payload The input data for updating the carrier rate.
 */
export const updateCarrierRate = graphql(`
  mutation UpdateCarrierRate($id: UUID!, $payload: CreateCarrierRateInput!) {
    tms {
      updateCarrierRate(id: $id, payload: $payload) {
        id
        serviceType
        origin
        destination
        rate
        unit
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Removes a carrier rate by its ID.
 * @param id The UUID of the carrier rate to remove.
 */
export const removeCarrierRate = graphql(`
  mutation RemoveCarrierRate($id: UUID!) {
    tms {
      removeCarrierRate(id: $id)
    }
  }
`);
